crates/metadata-store/src/omnipaxos/store.rs

```diff
@@ -0,0 +1,1232 @@
+            connection_manager,
+            request_rx,
+            join_cluster_rx,
+            metadata_writer,
+            storage_id,
+        )
+    }
+
+    fn become_active(self, omni_paxos_configuration: OmniPaxosConfiguration) -> Active {
+        let OmniPaxosMetadataStore {
+            connection_manager,
+            rocksdb_storage,
+            request_rx,
+            join_cluster_rx,
+            metadata_writer,
+            ..
+        } = self;
+
+        Active::new(
+            omni_paxos_configuration,
+            connection_manager,
+            rocksdb_storage,
+            request_rx,
+            join_cluster_rx,
+            metadata_writer,
+        )
+    }
+}
+
+enum Provisioned {
+    Active(Active),
+    Passive(Passive),
+}
+
+struct Active {
+    omni_paxos: Option<OmniPaxos>,
+    cluster_config: ClusterConfig,
+    members: HashMap<NodeId, StorageId>,
+
+    connection_manager: Arc<ArcSwapOption<ConnectionManager<OmniPaxosMessage>>>,
+    networking: Networking<OmniPaxosMessage>,
+    msg_rx: mpsc::Receiver<OmniPaxosMessage>,
+
+    own_member_id: MemberId,
+    is_leader: bool,
+
+    last_applied_index: usize,
+
+    pending_join_requests:
+        HashMap<MemberId, oneshot::Sender<Result<JoinClusterResponse, JoinClusterError>>>,
+
+    kv_storage: KvMemoryStorage,
+
+    request_rx: RequestReceiver,
+    join_cluster_rx: JoinClusterReceiver,
+
+    metadata_writer: Option<MetadataWriter>,
+}
+
+impl Active {
+    fn new(
+        omni_paxos_configuration: OmniPaxosConfiguration,
+        connection_manager: Arc<ArcSwapOption<ConnectionManager<OmniPaxosMessage>>>,
+        rocksdb_storage: RocksDbStorage<Request>,
+        request_rx: RequestReceiver,
+        join_cluster_rx: JoinClusterReceiver,
+        metadata_writer: Option<MetadataWriter>,
+    ) -> Active {
+        let own_member_id = omni_paxos_configuration.own_member_id;
+        let cluster_config = omni_paxos_configuration.cluster_config;
+        let members = omni_paxos_configuration.members;
+
+        let (router_tx, router_rx) = mpsc::channel(128);
+        let new_connection_manager = ConnectionManager::new(own_member_id.node_id, router_tx);
+        let mut networking = Networking::new(new_connection_manager.clone());
+
+        networking.register_address(
+            own_member_id.node_id,
+            Configuration::pinned().common.advertised_address.clone(),
+        );
+
+        let nodes_config = Metadata::with_current(|m| m.nodes_config_ref());
+
+        for node_id in &cluster_config.nodes {
+            if let Ok(node_config) = nodes_config.find_node_by_id(PlainNodeId::from(
+                u32::try_from(*node_id).expect("node is derived from PlainNodeId"),
+            )) {
+                networking.register_address(*node_id, node_config.address.clone());
+            }
+        }
+
+        // todo remove additional indirection from Arc
+        connection_manager.store(Some(Arc::new(new_connection_manager)));
```

Comment:
The `ConnectionManager` needs to know our own metadata peer id to validate that connections are set up correctly. Since we only know what our id will be once we have joined an OmniPaxos cluster, we need to instantiate it lazily. This is not super beautiful :-(